RHDEVDOCS-6360: Reviews for External secrets stores (#96694)
openshift/openshift-docs: modules/op-csi-secrets.adoc

          mountPath: "/mnt/secrets-store"
          readOnly: true
  volumes:
    - name: secrets-store-inline
      csi:
        driver: secrets-store.csi.k8s.io
        readOnly: true
        volumeAttributes:
          secretProviderClass: "vault-secret"
----
+
[NOTE]
====
The mount path that you specify in the pipeline or task definition overrides the path that you specify in the `SecretProviderClass` CR.
====

. In the `PipelineRun` or `TaskRun` CR, assign the service account that is authenticated with the secret store to the task:

** If you create a `PipelineRun` CR, use the `taskRunSpecs` section to assign the service account to the particular task. Do not assign the service account to the entire pipeline.
+
.Example `PipelineRun` CR
[source,yaml]
----
apiVersion: tekton.dev/v1
kind: PipelineRun
metadata:
  name: vault-secret-pipeline-run
  namespace: tekton-vault
spec:
  pipelineRef:
    name: vault-secret-pipeline
  taskRunSpecs:
    - pipelineTaskName: secret-task
      serviceAccountName: pipeline-sa
----

** If you create a `TaskRun` CR, use the `serviceAccountName` setting to assign the service account.
+
.Example `TaskRun` CR
[source,yaml]
----

> kind: TaskRun

Reviewer: please add `apiVersion: tekton.dev/v1`

+
.Example definition of the `pipeline-role-binding` role binding
[source,yaml]
----
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: pipeline-role-binding
  namespace: tekton-vault
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: Role
  name: pipeline-role
subjects:
- kind: ServiceAccount
  name: pipeline-sa
  namespace: tekton-vault
----

. Create a `SecretProviderClass` CR that defines the secrets that your pipeline or task consumes and the mount path and filename (key) for each of these secrets.
+
.Example definition of the `vault-secret` `SecretProviderClass` CR
[source,yaml]
----
apiVersion: secrets-store.csi.x-k8s.io/v1
kind: SecretProviderClass
metadata:
  name: vault-secret
  namespace: tekton-vault
spec:
  provider: vault
  parameters:
    vaultAddress: "http://vault.vault:8200"
    vaultSkipTLSVerify: "true"
    roleName: "my-role"
    objects: |
      - objectName: "demo-secret"
        secretPath: "secret/data/my-secret"
        secretKey: "username"
      - objectName: "demo-secret-pass"
        secretPath: "secret/data/my-secret"
        secretKey: "password"
----
+

> [NOTE]

Reviewer: ```suggestion
[WARNING]
```

Author: Fixed.

  name: secret-task
  namespace: tekton-vault
spec:
  steps:
    - name: use-secret
      image: registry.access.redhat.com/ubi8/ubi-minimal:latest
      script: |
        #!/bin/sh
        echo "Reading secrets from mounted volume..."
        echo "Username: $(cat /mnt/secrets-store/demo-secret)"
        echo "Password: $(cat /mnt/secrets-store/demo-secret-pass)"
      volumeMounts:
        - name: secrets-store-inline
          mountPath: "/mnt/secrets-store"
          readOnly: true
  volumes:
    - name: secrets-store-inline
      csi:
        driver: secrets-store.csi.k8s.io
        readOnly: true
        volumeAttributes:
          secretProviderClass: "vault-secret"
----
+
[NOTE]
====
The mount path that you specify in the pipeline or task definition overrides the path that you specify in the `SecretProviderClass` CR.
====

. In the `PipelineRun` or `TaskRun` CR, assign the service account that is authenticated with the secret store to the task:

** If you create a `PipelineRun` CR, use the `taskRunSpecs` section to assign the service account to the particular task. Do not assign the service account to the entire pipeline.
+
.Example `PipelineRun` CR
[source,yaml]
----

> apiVersion: tekton.dev/v1
> kind: Pipeline
> metadata:
>   name: vault-secret-pipeline
>   namespace: tekton-vault
> spec:
>   tasks:
>     - name: read-secret
>       taskRef:
>         name: secret-task
>       serviceAccountName: pipeline-sa

Reviewer: ```suggestion
apiVersion: tekton.dev/v1
kind: Pipeline
metadata:
  name: vault-secret-pipeline
  namespace: tekton-vault
spec:
  tasks:
    - name: read-secret
      taskRef:
        name: secret-task
      serviceAccountName: pipeline-sa
---
apiVersion: tekton.dev/v1
kind: PipelineRun
metadata:
  name: vault-secret-pipeline-run
  namespace: tekton-vault
spec:
  pipelineRef:
    name: vault-secret-pipeline
  taskRunSpecs:
    - pipelineTaskName: secret-task
      serviceAccountName: pipeline-sa
```

Reviewer: Both pipeline and pipelinerun need to be there in this case

Author: Fixed.

----

** If you create a `TaskRun` CR, use the `serviceAccountName` setting to assign the service account.
+
.Example `TaskRun` CR
[source,yaml]
----
apiVersion: tekton.dev/v1
kind: TaskRun
metadata:

>   name: vault-secret-task-run

Reviewer: ```suggestion
  name: vault-secret-task-run
  namespace: tekton-vault
```

====

. In the definition of the task that consumes the secrets, define and mount a volume that uses the CSI secrets store driver and specifies the name of the `SecretProviderClass` CR. To minimise security exposure of secrets, mount them in the definition of a task and not of an entire pipeline.
+
.Example definition of a task that consumes secrets
[source,yaml]
----
apiVersion: tekton.dev/v1
kind: Task
metadata:
  name: secret-task
  namespace: tekton-vault
spec:
  steps:
    - name: use-secret
      image: registry.access.redhat.com/ubi8/ubi-minimal:latest
      script: |
        #!/bin/sh
        echo "Reading secrets from mounted volume..."
        echo "Username: $(cat /mnt/secrets-store/demo-secret)"
        echo "Password: $(cat /mnt/secrets-store/demo-secret-pass)"
      volumeMounts:
        - name: secrets-store-inline
          mountPath: "/mnt/secrets-store"
          readOnly: true
  volumes:
    - name: secrets-store-inline
      csi:
        driver: secrets-store.csi.k8s.io
        readOnly: true
        volumeAttributes:
          secretProviderClass: "vault-secret"

> ----

Author: Fixed.

+
[NOTE]
====
The mount path that you specify in the pipeline or task definition overrides the path that you specify in the `SecretProviderClass` CR.
====

. In the `PipelineRun` or `TaskRun` CR, assign the service account that is authenticated with the secret store to the task:

> ** If you create a `PipelineRun` CR, use the `taskRunSpecs` section to assign the service account to the particular task. Do not assign the service account to the entire pipeline.

Author: Fixed as per discussion with @waveywaves.

+
.Example `PipelineRun` CR
[source,yaml]
----
apiVersion: tekton.dev/v1
kind: Pipeline
metadata:
  name: vault-secret-pipeline
  namespace: tekton-vault
spec:
  tasks:
    - name: read-secret
      taskRef:
        name: secret-task
      serviceAccountName: pipeline-sa
---
apiVersion: tekton.dev/v1
kind: PipelineRun
metadata:
  name: vault-secret-pipeline-run
  namespace: tekton-vault
spec:
  pipelineRef:
    name: vault-secret-pipeline
  taskRunSpecs:

>     - pipelineTaskName: secret-task

Reviewer: ```suggestion
    - pipelineTaskName: read-secret
```

Author: Fixed.

      serviceAccountName: pipeline-sa
----

** If you create a `TaskRun` CR, use the `serviceAccountName` setting to assign the service account.
+
.Example `TaskRun` CR
[source,yaml]
----
apiVersion: tekton.dev/v1
kind: TaskRun
metadata:
  name: vault-secret-task-run
  namespace: tekton-vault
spec:
  taskRef:
    name: secret-task

>   serviceAccountName: pipeline-sa

Reviewer 2: failed to create task run pod "vault-secret-task-run": pods "vault-secret-task-run-pod" is forbidden: vault-secret-task-run-pod uses an inline volume provided by CSIDriver secrets-store.csi.k8s.io and namespace tekton-vault has a pod security enforce level that is lower than privileged. Maybe missing or invalid Task tekton-vault/secret-task

====

* You configured authentication with the secret store for an {OCP} service account, for example, `pipeline-sa`. For information about configuring Kubernetes authentication with Hashicorp Vault, see link:https://developer.hashicorp.com/vault/docs/auth/kubernetes[Kubernetes auth method] in the Hashicorp documentation.
+
[NOTE]
====
You must create the service account in the namespace in which you create pipelines and other Custom Resources (CRs) for {pipelines-shortname}.
====

* Ensure that the Secrets Store CSI Driver service account has the required cluster-wide permissions to create service account tokens in all relevant namespaces. Without these permissions, pods will fail to start and you may encounter the following error:
+
[source,terminal]
----
User "system:serviceaccount:csi-driver:secrets-store-csi-driver" cannot create resource "serviceaccounts/token" in API group "" in the namespace "tekton-vault"
----

.Procedure

> . To configure the namespace to allow privileged pod security for CSI volumes, run the following command:

Reviewer 3: Suggestion: Just a reword.
```suggestion
. To allow the CSI driver to mount secrets volumes, configure the namespace to allow privileged pod security:
```
